File: backend/app/schemas/plugins.py
"""
Plugin Architecture Schemas

Pydantic schemas for plugin manifests, lifecycle, hooks, and registry.
"""

from typing import Any, Optional, Union
from pydantic import BaseModel, Field
from enum import Enum
from datetime import datetime


# Enums

class PluginType(str, Enum):
    """Types of plugins"""
    CONNECTOR = "connector"  # Data source connectors
    VISUALIZATION = "visualization"  # Custom chart types
    TRANSFORM = "transform"  # Data transformations
    EXPORT = "export"  # Export formats
    AUTH = "auth"  # Authentication providers
    THEME = "theme"  # UI themes
    WIDGET = "widget"  # Dashboard widgets
    INTEGRATION = "integration"  # Third-party integrations
    AI = "ai"  # AI/ML extensions


class PluginStatus(str, Enum):
    """Plugin status"""
    AVAILABLE = "available"  # In registry, not installed
    INSTALLED = "installed"  # Installed but not enabled
    ENABLED = "enabled"  # Active and running
    DISABLED = "disabled"  # Installed but disabled
    ERROR = "error"  # Failed to load
    UPDATING = "updating"  # Being updated
    DEPRECATED = "deprecated"  # Marked for removal


class HookType(str, Enum):
    """Plugin hook types"""
    # Data hooks
    BEFORE_QUERY = "before_query"
    AFTER_QUERY = "after_query"
    BEFORE_TRANSFORM = "before_transform"
    AFTER_TRANSFORM = "after_transform"

    # UI hooks
    DASHBOARD_RENDER = "dashboard_render"
    CHART_RENDER = "chart_render"
    SIDEBAR_MENU = "sidebar_menu"
    TOOLBAR_ACTION = "toolbar_action"
    CONTEXT_MENU = "context_menu"

    # Auth hooks
    BEFORE_AUTH = "before_auth"
    AFTER_AUTH = "after_auth"

    # Export hooks
    BEFORE_EXPORT = "before_export"
    AFTER_EXPORT = "after_export"

    # Lifecycle hooks
    ON_INSTALL = "on_install"
    ON_ENABLE = "on_enable"
    ON_DISABLE = "on_disable"
    ON_UNINSTALL = "on_uninstall"
    ON_UPDATE = "on_update"


class PermissionScope(str, Enum):
    """Plugin permission scopes"""
    READ_DATA = "read_data"
    WRITE_DATA = "write_data"
    READ_CONFIG = "read_config"
    WRITE_CONFIG = "write_config"
    NETWORK = "network"
    STORAGE = "storage"
    UI = "ui"
    AUTH = "auth"
    ADMIN = "admin"


# Manifest Models

class PluginAuthor(BaseModel):
    """Plugin author information"""
    name: str
    email: Optional[str] = None
    url: Optional[str] = None


class PluginDependency(BaseModel):
    """Plugin dependency"""
    plugin_id: str
    version: str  # Semver range, e.g., "^1.0.0"
    optional: bool = False


class PluginAsset(BaseModel):
    """Plugin asset (JS, CSS, images)"""
    type: str  # js, css, image, font
    path: str
    load_on: str = "enable"  # enable, dashboard, chart, etc.
    priority: int = 0


class PluginConfig(BaseModel):
    """Plugin configuration schema"""
    key: str
    type: str  # string, number, boolean, select, json
    label: str
    description: Optional[str] = None
    default: Optional[Any] = None
    required: bool = False
    options: list[dict[str, Any]] = Field(default_factory=list)  # For select type
    validation: Optional[dict[str, Any]] = None


class PluginHook(BaseModel):
    """Plugin hook registration"""
    type: HookType
    handler: str  # Function name or path
    priority: int = 10  # Lower = runs first
    async_: bool = Field(False, alias="async")
    conditions: dict[str, Any] = Field(default_factory=dict)


class PluginEndpoint(BaseModel):
    """Custom API endpoint provided by plugin"""
    path: str
    method: str
    handler: str
    description: Optional[str] = None
    requires_auth: bool = True


class PluginManifest(BaseModel):
    """Plugin manifest (plugin.json)"""
    # Required fields
    id: str  # Unique identifier, e.g., "bheem-connector-snowflake"
    name: str
    version: str  # Semver
    description: str
    type: PluginType

    # Author info
    author: PluginAuthor
    license: str = "MIT"
    homepage: Optional[str] = None
    repository: Optional[str] = None

    # Compatibility
    min_app_version: str = "1.0.0"
    max_app_version: Optional[str] = None

    # Entry points
    main: Optional[str] = None  # Server-side entry
    client: Optional[str] = None  # Client-side entry

    # Dependencies
    dependencies: list[PluginDependency] = Field(default_factory=list)
    peer_dependencies: list[str] = Field(default_factory=list)

    # Permissions
    permissions: list[PermissionScope] = Field(default_factory=list)

    # Configuration
    config_schema: list[PluginConfig] = Field(default_factory=list)

    # Hooks
    hooks: list[PluginHook] = Field(default_factory=list)

    # API endpoints
    endpoints: list[PluginEndpoint] = Field(default_factory=list)

    # Assets
    assets: list[PluginAsset] = Field(default_factory=list)

    # UI contributions
    sidebar_items: list[dict[str, Any]] = Field(default_factory=list)
    settings_pages: list[dict[str, Any]] = Field(default_factory=list)

    # Keywords for search
    keywords: list[str] = Field(default_factory=list)

    # Metadata
    icon: Optional[str] = None
    banner: Optional[str] = None
    screenshots: list[str] = Field(default_factory=list)
    changelog: Optional[str] = None


# Plugin Instance Models

class PluginInstance(BaseModel):
    """Installed plugin instance"""
    id: str
    manifest: PluginManifest
    status: PluginStatus = PluginStatus.INSTALLED

    # Installation info
    installed_at: datetime
    installed_by: str
    updated_at: Optional[datetime] = None
    version_installed: str

    # Configuration
    config: dict[str, Any] = Field(default_factory=dict)

    # Runtime
    enabled_at: Optional[datetime] = None
    disabled_at: Optional[datetime] = None
    last_error: Optional[str] = None
    error_count: int = 0

    # Usage
    load_time_ms: Optional[int] = None
    memory_usage_bytes: Optional[int] = None

    # Workspace scoping
    workspace_id: Optional[str] = None  # None = global
    enabled_workspaces: list[str] = Field(default_factory=list)


class PluginInstall(BaseModel):
    """Request to install a plugin"""
    plugin_id: str  # From registry
    version: Optional[str] = None  # None = latest
    config: dict[str, Any] = Field(default_factory=dict)
    workspace_id: Optional[str] = None
    enable_after_install: bool = True


class PluginUpdate(BaseModel):
    """Request to update a plugin"""
    config: Optional[dict[str, Any]] = None
    enabled_workspaces: Optional[list[str]] = None


# Registry Models

class RegistryPlugin(BaseModel):
    """Plugin in the registry"""
    id: str
    manifest: PluginManifest

    # Registry metadata
    downloads: int = 0
    rating: float = 0.0
    reviews_count: int = 0
    verified: bool = False
    featured: bool = False

    # Versions
    versions: list[str] = Field(default_factory=list)
    latest_version: str

    # Timestamps
    published_at: datetime
    updated_at: Optional[datetime] = None


class RegistrySearchResult(BaseModel):
    """Registry search results"""
    plugins: list[RegistryPlugin]
    total: int
    page: int
    page_size: int
    has_more: bool


class RegistrySearchQuery(BaseModel):
    """Registry search query"""
    query: Optional[str] = None
    type: Optional[PluginType] = None
    keywords: list[str] = Field(default_factory=list)
    verified_only: bool = False
    sort_by: str = "downloads"  # downloads, rating, updated, name
    page: int = 1
    page_size: int = 20


# Hook Execution Models

class HookContext(BaseModel):
    """Context passed to hook handlers"""
    hook_type: HookType
    plugin_id: str
    user_id: Optional[str] = None
    workspace_id: Optional[str] = None
    resource_type: Optional[str] = None
    resource_id: Optional[str] = None
    data: dict[str, Any] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)


class HookResult(BaseModel):
    """Result from hook execution"""
    plugin_id: str
    hook_type: HookType
    success: bool
    data: Optional[dict[str, Any]] = None
    error: Optional[str] = None
    execution_time_ms: int = 0


class HookChainResult(BaseModel):
    """Result from executing a chain of hooks"""
    hook_type: HookType
    results: list[HookResult]
    final_data: dict[str, Any]
    total_time_ms: int
    any_failed: bool


# Event Models

class PluginEvent(BaseModel):
    """Plugin lifecycle event"""
    id: str
    plugin_id: str
    event_type: str  # installed, enabled, disabled, error, etc.
    timestamp: datetime
    user_id: Optional[str] = None
    details: dict[str, Any] = Field(default_factory=dict)


# Response Models

class PluginListResponse(BaseModel):
    """List of plugins response"""
    plugins: list[PluginInstance]
    total: int


class PluginStatsResponse(BaseModel):
    """Plugin statistics"""
    total_installed: int
    total_enabled: int
    by_type: dict[str, int]
    by_status: dict[str, int]
    total_hooks_registered: int
    recent_errors: list[dict[str, Any]]


# Constants

OFFICIAL_REGISTRY_URL = "https://plugins.dataviz.bheemkodee.com"

DEFAULT_PLUGIN_CONFIG = {
    "enabled": True,
    "log_level": "info",
    "timeout_ms": 5000,
}

PLUGIN_TYPE_ICONS = {
    PluginType.CONNECTOR: "database",
    PluginType.VISUALIZATION: "bar-chart",
    PluginType.TRANSFORM: "git-branch",
    PluginType.EXPORT: "download",
    PluginType.AUTH: "lock",
    PluginType.THEME: "palette",
    PluginType.WIDGET: "layout",
    PluginType.INTEGRATION: "plug",
    PluginType.AI: "brain",
}

PLUGIN_TYPE_LABELS = {
    PluginType.CONNECTOR: "Data Connector",
    PluginType.VISUALIZATION: "Visualization",
    PluginType.TRANSFORM: "Transform",
    PluginType.EXPORT: "Export Format",
    PluginType.AUTH: "Authentication",
    PluginType.THEME: "Theme",
    PluginType.WIDGET: "Widget",
    PluginType.INTEGRATION: "Integration",
    PluginType.AI: "AI Extension",
}


# Helper Functions

def validate_semver(version: str) -> bool:
    """Validate semver version string."""
    import re
    pattern = r'^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$'
    return bool(re.match(pattern, version))


def compare_versions(v1: str, v2: str) -> int:
    """Compare two semver versions. Returns -1, 0, or 1."""
    def parse(v):
        parts = v.split('-')[0].split('.')
        return [int(p) for p in parts]

    p1, p2 = parse(v1), parse(v2)

    for a, b in zip(p1, p2):
        if a < b:
            return -1
        if a > b:
            return 1

    return 0


def is_compatible(plugin_version: str, min_version: str, max_version: Optional[str] = None) -> bool:
    """Check if plugin is compatible with app version."""
    if compare_versions(plugin_version, min_version) < 0:
        return False

    if max_version and compare_versions(plugin_version, max_version) > 0:
        return False

    return True


def generate_plugin_id(author: str, name: str) -> str:
    """Generate a plugin ID from author and name."""
    import re
    author_slug = re.sub(r'[^a-z0-9-]', '', author.lower().replace(' ', '-'))
    name_slug = re.sub(r'[^a-z0-9-]', '', name.lower().replace(' ', '-'))
    return f"{author_slug}-{name_slug}"

File: frontend/src/types/plugins.ts
/**
 * Plugin Architecture Types
 *
 * TypeScript types for plugin manifests, lifecycle, hooks, and registry.
 */

// Enums

export enum PluginType {
  CONNECTOR = 'connector',
  VISUALIZATION = 'visualization',
  TRANSFORM = 'transform',
  EXPORT = 'export',
  AUTH = 'auth',
  THEME = 'theme',
  WIDGET = 'widget',
  INTEGRATION = 'integration',
  AI = 'ai',
}

export enum PluginStatus {
  AVAILABLE = 'available',
  INSTALLED = 'installed',
  ENABLED = 'enabled',
  DISABLED = 'disabled',
  ERROR = 'error',
  UPDATING = 'updating',
  DEPRECATED = 'deprecated',
}

export enum HookType {
  BEFORE_QUERY = 'before_query',
  AFTER_QUERY = 'after_query',
  BEFORE_TRANSFORM = 'before_transform',
  AFTER_TRANSFORM = 'after_transform',
  DASHBOARD_RENDER = 'dashboard_render',
  CHART_RENDER = 'chart_render',
  SIDEBAR_MENU = 'sidebar_menu',
  TOOLBAR_ACTION = 'toolbar_action',
  CONTEXT_MENU = 'context_menu',
  BEFORE_AUTH = 'before_auth',
  AFTER_AUTH = 'after_auth',
  BEFORE_EXPORT = 'before_export',
  AFTER_EXPORT = 'after_export',
  ON_INSTALL = 'on_install',
  ON_ENABLE = 'on_enable',
  ON_DISABLE = 'on_disable',
  ON_UNINSTALL = 'on_uninstall',
  ON_UPDATE = 'on_update',
}

export enum PermissionScope {
  READ_DATA = 'read_data',
  WRITE_DATA = 'write_data',
  READ_CONFIG = 'read_config',
  WRITE_CONFIG = 'write_config',
  NETWORK = 'network',
  STORAGE = 'storage',
  UI = 'ui',
  AUTH = 'auth',
  ADMIN = 'admin',
}

// Manifest Types

export interface PluginAuthor {
  name: string;
  email?: string | null;
  url?: string | null;
}

export interface PluginDependency {
  plugin_id: string;
  version: string;
  optional: boolean;
}

export interface PluginAsset {
  type: string;
  path: string;
  load_on: string;
  priority: number;
}

export interface PluginConfig {
  key: string;
  type: string;
  label: string;
  description?: string | null;
  default?: unknown;
  required: boolean;
  options: Array<Record<string, unknown>>;
  validation?: Record<string, unknown> | null;
}

export interface PluginHook {
  type: HookType;
  handler: string;
  priority: number;
  async: boolean;
  conditions: Record<string, unknown>;
}

export interface PluginEndpoint {
  path: string;
  method: string;
  handler: string;
  description?: string | null;
  requires_auth: boolean;
}

export interface PluginManifest {
  id: string;
  name: string;
  version: string;
  description: string;
  type: PluginType;
  author: PluginAuthor;
  license: string;
  homepage?: string | null;
  repository?: string | null;
  min_app_version: string;
  max_app_version?: string | null;
  main?: string | null;
  client?: string | null;
  dependencies: PluginDependency[];
  peer_dependencies: string[];
  permissions: PermissionScope[];
  config_schema: PluginConfig[];
  hooks: PluginHook[];
  endpoints: PluginEndpoint[];
  assets: PluginAsset[];
  sidebar_items: Array<Record<string, unknown>>;
  settings_pages: Array<Record<string, unknown>>;
  keywords: string[];
  icon?: string | null;
  banner?: string | null;
  screenshots: string[];
  changelog?: string | null;
}

// Plugin Instance Types

export interface PluginInstance {
  id: string;
  manifest: PluginManifest;
  status: PluginStatus;
  installed_at: string;
  installed_by: string;
  updated_at?: string | null;
  version_installed: string;
  config: Record<string, unknown>;
  enabled_at?: string | null;
  disabled_at?: string | null;
  last_error?: string | null;
  error_count: number;
  load_time_ms?: number | null;
  memory_usage_bytes?: number | null;
  workspace_id?: string | null;
  enabled_workspaces: string[];
}

export interface PluginInstall {
  plugin_id: string;
  version?: string | null;
  config?: Record<string, unknown>;
  workspace_id?: string | null;
  enable_after_install?: boolean;
}

export interface PluginUpdate {
  config?: Record<string, unknown> | null;
  enabled_workspaces?: string[] | null;
}

// Registry Types

export interface RegistryPlugin {
  id: string;
  manifest: PluginManifest;
  downloads: number;
  rating: number;
  reviews_count: number;
  verified: boolean;
  featured: boolean;
  versions: string[];
  latest_version: string;
  published_at: string;
  updated_at?: string | null;
}

export interface RegistrySearchResult {
  plugins: RegistryPlugin[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface RegistrySearchQuery {
  query?: string | null;
  type?: PluginType | null;
  keywords?: string[];
  verified_only?: boolean;
  sort_by?: string;
  page?: number;
  page_size?: number;
}

// Hook Types

export interface HookContext {
  hook_type: HookType;
  plugin_id: string;
  user_id?: string | null;
  workspace_id?: string | null;
  resource_type?: string | null;
  resource_id?: string | null;
  data: Record<string, unknown>;
  metadata: Record<string, unknown>;
}

export interface HookResult {
  plugin_id: string;
  hook_type: HookType;
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
  execution_time_ms: number;
}

export interface HookChainResult {
  hook_type: HookType;
  results: HookResult[];
  final_data: Record<string, unknown>;
  total_time_ms: number;
  any_failed: boolean;
}

// Event Types

export interface PluginEvent {
  id: string;
  plugin_id: string;
  event_type: string;
  timestamp: string;
  user_id?: string | null;
  details: Record<string, unknown>;
}

// Response Types

export interface PluginListResponse {
  plugins: PluginInstance[];
  total: number;
}

export interface PluginStatsResponse {
  total_installed: number;
  total_enabled: number;
  by_type: Record<string, number>;
  by_status: Record<string, number>;
  total_hooks_registered: number;
  recent_errors: Array<{
    plugin_id: string;
    error: string;
    timestamp: string;
  }>;
}

export interface UpdateCheckResult {
  updates_available: number;
  plugins: Array<{
    plugin_id: string;
    current_version: string;
    latest_version: string;
    plugin_name: string;
  }>;
}

// Constants

export const PLUGIN_TYPE_LABELS: Record<PluginType, string> = {
  [PluginType.CONNECTOR]: 'Data Connector',
  [PluginType.VISUALIZATION]: 'Visualization',
  [PluginType.TRANSFORM]: 'Transform',
  [PluginType.EXPORT]: 'Export Format',
  [PluginType.AUTH]: 'Authentication',
  [PluginType.THEME]: 'Theme',
  [PluginType.WIDGET]: 'Widget',
  [PluginType.INTEGRATION]: 'Integration',
  [PluginType.AI]: 'AI Extension',
};

export const PLUGIN_TYPE_ICONS: Record<PluginType, string> = {
  [PluginType.CONNECTOR]: 'database',
  [PluginType.VISUALIZATION]: 'bar-chart',
  [PluginType.TRANSFORM]: 'git-branch',
  [PluginType.EXPORT]: 'download',
  [PluginType.AUTH]: 'lock',
  [PluginType.THEME]: 'palette',
  [PluginType.WIDGET]: 'layout',
  [PluginType.INTEGRATION]: 'plug',
  [PluginType.AI]: 'brain',
};

export const PLUGIN_STATUS_LABELS: Record<PluginStatus, string> = {
  [PluginStatus.AVAILABLE]: 'Available',
  [PluginStatus.INSTALLED]: 'Installed',
  [PluginStatus.ENABLED]: 'Enabled',
  [PluginStatus.DISABLED]: 'Disabled',
  [PluginStatus.ERROR]: 'Error',
  [PluginStatus.UPDATING]: 'Updating',
  [PluginStatus.DEPRECATED]: 'Deprecated',
};

export const PLUGIN_STATUS_COLORS: Record<PluginStatus, string> = {
  [PluginStatus.AVAILABLE]: 'gray',
  [PluginStatus.INSTALLED]: 'blue',
  [PluginStatus.ENABLED]: 'green',
  [PluginStatus.DISABLED]: 'yellow',
  [PluginStatus.ERROR]: 'red',
  [PluginStatus.UPDATING]: 'blue',
  [PluginStatus.DEPRECATED]: 'orange',
};

export const PERMISSION_LABELS: Record<PermissionScope, string> = {
  [PermissionScope.READ_DATA]: 'Read Data',
  [PermissionScope.WRITE_DATA]: 'Write Data',
  [PermissionScope.READ_CONFIG]: 'Read Configuration',
  [PermissionScope.WRITE_CONFIG]: 'Write Configuration',
  [PermissionScope.NETWORK]: 'Network Access',
  [PermissionScope.STORAGE]: 'Storage Access',
  [PermissionScope.UI]: 'UI Modifications',
  [PermissionScope.AUTH]: 'Authentication',
  [PermissionScope.ADMIN]: 'Admin Access',
};

// Helper Functions

export function isPluginEnabled(plugin: PluginInstance): boolean {
  return plugin.status === PluginStatus.ENABLED;
}

export function isPluginInstalled(plugin: PluginInstance): boolean {
  return [
    PluginStatus.INSTALLED,
    PluginStatus.ENABLED,
    PluginStatus.DISABLED,
  ].includes(plugin.status);
}

export function hasUpdate(installed: PluginInstance, registry: RegistryPlugin): boolean {
  return compareVersions(registry.latest_version, installed.version_installed) > 0;
}

export function compareVersions(v1: string, v2: string): number {
  const parse = (v: string) => v.split('-')[0].split('.').map(Number);
  const p1 = parse(v1);
  const p2 = parse(v2);

  for (let i = 0; i < Math.max(p1.length, p2.length); i++) {
    const a = p1[i] || 0;
    const b = p2[i] || 0;
    if (a < b) return -1;
    if (a > b) return 1;
  }
  return 0;
}

export function formatDownloads(count: number): string {
  if (count >= 1000000) return `${(count / 1000000).toFixed(1)}M`;
  if (count >= 1000) return `${(count / 1000).toFixed(1)}K`;
  return count.toString();
}

export function formatRating(rating: number): string {
  return rating.toFixed(1);
}

export function getStarRating(rating: number): number[] {
  const stars = [];
  for (let i = 1; i <= 5; i++) {
    if (rating >= i) {
      stars.push(1); // Full star
    } else if (rating >= i - 0.5) {
      stars.push(0.5); // Half star
    } else {
      stars.push(0); // Empty star
    }
  }
  return stars;
}

// Config Form Helpers

export interface ConfigFormField {
  key: string;
  type: 'string' | 'number' | 'boolean' | 'select' | 'json';
  label: string;
  description?: string;
  defaultValue?: unknown;
  required: boolean;
  options?: Array<{ value: unknown; label: string }>;
  validation?: {
    min?: number;
    max?: number;
    pattern?: string;
    message?: string;
  };
}

export function manifestConfigToFormFields(schema: PluginConfig[]): ConfigFormField[] {
  return schema.map((config) => ({
    key: config.key,
    type: config.type as ConfigFormField['type'],
    label: config.label,
    description: config.description || undefined,
    defaultValue: config.default,
    required: config.required,
    options: config.options.map((opt) => ({
      value: opt.value,
      label: opt.label as string,
    })),
    validation: config.validation as ConfigFormField['validation'],
  }));
}

export function validateConfig(
  config: Record<string, unknown>,
  schema: PluginConfig[]
): { valid: boolean; errors: Record<string, string> } {
  const errors: Record<string, string> = {};

  for (const field of schema) {
    const value = config[field.key];

    if (field.required && (value === undefined || value === null || value === '')) {
      errors[field.key] = `${field.label} is required`;
      continue;
    }

    if (value !== undefined && field.validation) {
      const val = field.validation as Record<string, unknown>;
      if (field.type === 'number') {
        const num = Number(value);
        if (val.min !== undefined && num < (val.min as number)) {
          errors[field.key] = `Must be at least ${val.min}`;
        }
        if (val.max !== undefined && num > (val.max as number)) {
          errors[field.key] = `Must be at most ${val.max}`;
        }
      }
      if (field.type === 'string' && val.pattern) {
        const regex = new RegExp(val.pattern as string);
        if (!regex.test(String(value))) {
          errors[field.key] = (val.message as string) || 'Invalid format';
        }
      }
    }
  }

  return { valid: Object.keys(errors).length === 0, errors };
}

// Plugin State Management

export interface PluginManagerState {
  installedPlugins: PluginInstance[];
  registryPlugins: RegistryPlugin[];
  selectedPlugin: PluginInstance | RegistryPlugin | null;
  isLoading: boolean;
  error: string | null;
  searchQuery: string;
  filterType: PluginType | null;
  filterStatus: PluginStatus | null;
  sortBy: 'name' | 'downloads' | 'rating' | 'updated';
  stats: PluginStatsResponse | null;
  updatesAvailable: UpdateCheckResult | null;
}

export function createInitialPluginState(): PluginManagerState {
  return {
    installedPlugins: [],
    registryPlugins: [],
    selectedPlugin: null,
    isLoading: false,
    error: null,
    searchQuery: '',
    filterType: null,
    filterStatus: null,
    sortBy: 'downloads',
    stats: null,
    updatesAvailable: null,
  };
}
